### PR TITLE
Scheduling with alfred and new syntax.

### PR DESCRIPTION
**Pull request by MrYawe**
### Example
**launch_cron.js**
```js
require("./client");

const CronWorkflow = require("./Workflows/CronWorkflow");

await new CronWorkflow().schedule("* * * * *")
```

Same thing for tasks.

### CODE REVIEW

File: src/async/Client.js [hunk 1]
```diff
@@ -113,6 +115,14 @@ module.exports = class Client {
     return `${host}${path}`;
   }
 
+  getAlfredUrl() {
+    const host = process.env.ZENATON_ALFRED_URL
+      ? process.env.ZENATON_ALFRED_URL
+      : ZENATON_ALFRED_URL;
+
+    return `${host}`;
```
Comment (jalric) [inline src/async/Client.js]: ?! 🤔like the consistency with other getters, but look like weird.

Comment (MrYawe) [reply src/async/Client.js]: I forgot to remove the template syntax thanks

File: src/async/Engine/Engine.js [hunk 2]
```diff
@@ -73,9 +73,8 @@ module.exports = class Engine {
           ? this.client.startWorkflow(job)
           : this.client.startTask(job);
 
-        await handler;
-
-        return undefined;
+        const res = await handler;
+        return res;
```
Comment (jalric) [inline src/async/Engine/Engine.js]: You can just return handler: `return handler`

Comment (MrYawe) [reply src/async/Engine/Engine.js]: @jalric 
That what I thought but I have the following warning if I do so:
https://eslint.org/docs/rules/no-return-await

File: src/async/Services/GraphQL.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+function getError(err) {
+  // Validation errors
+  if (err.response && err.response.errors && err.response.errors.length > 0) {
+    const message = err.response.errors
+      .map((graphqlError) => {
+        const path = graphqlError.path ? `(${graphqlError.path}) ` : "";
+        const errorMessage = graphqlError.message
+          ? graphqlError.message
+          : "Unknown error";
```
Comment (jalric) [inline src/async/Services/GraphQL.js]: a little JS trick:
You can write here `const errorMessage = graphqlError.message || "Unknown error"`

File: src/async/Engine/Engine.js (hunk 2 above)
Comment (jalric) [reply src/async/Engine/Engine.js]: no need to return await, just return the handler.

File: src/async/Services/GraphQL.js [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+      headers: {
+        Accept: "application/json",
+        "Content-Type": "application/json",
+        "app-id": credentials.appId,
+        "api-token": credentials.apiToken,
+      },
+    });
+
+    const response = await graphQLClient.request(query, variables);
+    return response;
```
Comment (jalric) [inline src/async/Services/GraphQL.js]: Same as above

**Review by jalric — APPROVED** [review]

**Comment by geomagilles** [comment]
@MrYawe let's talk about the output, I'm not sure to understand it fully.
- id: what's this?
- intentId: this should not the exposed naming (scheduleId)
- targetType: why not a type in target? (and btw always use lowercase 'workflow' - we have to be consistent) 
- why a name?
- it seems to me that canonicalName should be "CronWorkflow"
- a date is probably missing "scheduledAt"

**Comment by MrYawe** [comment]
@geomagilles 
> id: what's this?

it's the id of the resource we just created (a `schedule`)

> intentId: this should not the exposed naming (scheduleId)

I'm not sure to understand but today all requests made by the lib to the agent have an `intent_id`. I think we should do the same between the lib and alfred. You just want to "hide" this field to the client ? 

> targetType: why not a type in target? (and btw always use lowercase 'workflow' - we have to be consistent)

👍 

> why a name?

I think it may be useful to be able to choose the name of a `schedule` as an option (if you have multiple `schedule` for the same workflow for example). We can't pass this option from the lib today (we will wait for the new syntax ?) but this option is available and ready server side. Today the name of a `schedule` is an optional parameter and by default it's the name of the target (workflow or task).

> it seems to me that canonicalName should be "CronWorkflow"

👍 

> a date is probably missing "scheduledAt"

yes. I will add the `inserted_at` and `updated_at` in the GraphQL response.

**Review by pylebecq — APPROVED** [review]
Seems good to me.

The only question I have is why do you uppercase the programming language?

**Comment by MrYawe** [comment]
@pylebecq 
Because `programming_language` is a [GraphQL enum type ](https://graphql.org/learn/schema/#enumeration-types) so it must be an uppercased string.

**Pull request merged**